go/proxyd/daisychain.go

```diff
@@ -0,0 +1,599 @@
+		writeRPCRes(ctx, w, backendRes)
+		return
+	}
+
+	// If the epoch is not set, default to the latest
+	if argument.Epoch == nil {
+		argument.Epoch = &latestEpoch
+	}
+
+	url := ""
+	switch *argument.Epoch {
+	case 6:
+		url = s.epoch6RPCURL
+	case 5:
+		url = s.epoch5RPCURL
+	case 4:
+		url = s.epoch4RPCURL
+	case 3:
+		url = s.epoch3RPCURL
+	case 2:
+		url = s.epoch2RPCURL
+	case 1:
+		url = s.epoch1RPCURL
+	default:
+		writeRPCError(ctx, w, nil, errors.New("bad epoch"))
+		return
+	}
+
+	// TODO: delete this so a url with a key doesn't leak
+	log.Info("Sending rpc req", "url", url)
+
+	if url == "" {
+		writeRPCError(ctx, w, nil, errors.New("epoch not configured"))
+		return
+	}
+
+	// There should never be an empty params by this point
+	raw, err := json.Marshal(values[0 : len(values)-1])
+	if err != nil {
+		writeRPCError(ctx, w, nil, errors.New("cannot serialize json"))
+		return
+	}
+
+	req.Params = raw
+	backendRes := s.handleSingleRPC(ctx, url, req)
+	writeRPCRes(ctx, w, backendRes)
+}
+
+func (s *DaisyChainServer) RPCListenAndServe(host string, port int) error {
+	hdlr := mux.NewRouter()
+	hdlr.HandleFunc("/healthz", s.HandleHealthz).Methods("GET")
+	hdlr.HandleFunc("/", s.HandleRPC).Methods("POST")
+	c := cors.New(cors.Options{
+		AllowedOrigins: []string{"*"},
+	})
+	addr := fmt.Sprintf("%s:%d", host, port)
+	s.rpcServer = &http.Server{
+		Handler: instrumentedHdlr(c.Handler(hdlr)),
+		Addr:    addr,
+	}
+	log.Info("starting HTTP server", "addr", addr)
+	return s.rpcServer.ListenAndServe()
+}
+
+func (s *DaisyChainServer) WSListenAndServe(host string, port int) error {
+	hdlr := mux.NewRouter()
+	hdlr.HandleFunc("/healthz", s.HandleHealthz).Methods("GET")
+	// TODO: fix
+	//hdlr.HandleFunc("/", s.HandleWS)
+	c := cors.New(cors.Options{
+		AllowedOrigins: []string{"*"},
+	})
+	addr := fmt.Sprintf("%s:%d", host, port)
+	s.wsServer = &http.Server{
+		Handler: instrumentedHdlr(c.Handler(hdlr)),
+		Addr:    addr,
+	}
+	log.Info("starting WS server", "addr", addr)
+	return s.wsServer.ListenAndServe()
+}
+
+func (s *DaisyChainServer) Shutdown() {
+	if s.rpcServer != nil {
+		s.rpcServer.Shutdown(context.Background())
+	}
+	if s.wsServer != nil {
+		s.wsServer.Shutdown(context.Background())
+	}
+}
+
+func (s *DaisyChainServer) HandleHealthz(w http.ResponseWriter, r *http.Request) {
+	w.Write([]byte("OK"))
+}
+
+func (s *DaisyChainServer) populateContext(w http.ResponseWriter, r *http.Request) context.Context {
+	xff := r.Header.Get("X-Forwarded-For")
+	if xff == "" {
+		ipPort := strings.Split(r.RemoteAddr, ":")
+		if len(ipPort) == 2 {
+			xff = ipPort[0]
+		}
+	}
+
+	ctx := context.WithValue(r.Context(), ContextKeyXForwardedFor, xff)
+	return context.WithValue(
+		ctx,
+		ContextKeyReqID,
+		randStr(10),
+	)
+}
+
+func (s *DaisyChainServer) handleSingleRPC(ctx context.Context, url string, req *RPCReq) *RPCRes {
+	if url == "" {
+		return NewRPCErrorRes(nil, errors.New("no backend url"))
+	}
+
+	if err := ValidateRPCReq(req); err != nil {
```

Comment:
Can we use the proxyd backend objects here? For example, each epoch is a proxd `Backend`. This way we can do rate limiting, collect error metrics for each backend, etc.

Reply:
Each epoch is now a `Backend`